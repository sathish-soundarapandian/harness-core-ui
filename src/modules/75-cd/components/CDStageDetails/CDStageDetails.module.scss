/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

.container {
  display: flex;
  align-items: center;
  height: 100%;
}
.main {
  padding: 0 var(--spacing-6);
  display: grid;
  grid-template-columns: 1fr 1fr;
  width: 100%;
  align-items: start;
  &.threeSections {
    grid-template-columns: 1fr 1fr 1fr;
  }

  .title {
    font-size: var(--font-size-small);
    font-weight: 500;
    margin-bottom: var(--spacing-1);
  }

  .serviceName {
    display: inline-block;
    cursor: pointer;
  }

  .serviceWrapper {
    display: inline-block;
  }
}

ul.values {
  padding: 0;
  margin: 0;
  list-style-type: none;

  li,
  .git-ops-env-text {
    font-size: var(--font-size-small) !important;
    color: var(--primary-9);
    line-height: 1.33 !important;
  }
}

.gitOpsAppsLi {
  display: flex;
}

.stageItemDetails {
  max-width: 250px;
}
