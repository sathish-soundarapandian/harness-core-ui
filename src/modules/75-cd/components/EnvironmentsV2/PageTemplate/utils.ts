/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { useFeatureFlags } from '@common/hooks/useFeatureFlag'
import type { FilterProperties } from 'services/cd-ng'
import type { UseQueryParamsOptions } from '@common/hooks/useQueryParams';
import { useQueryParamsOptions } from '@common/hooks/useQueryParams'
import { COMMON_DEFAULT_PAGE_SIZE } from '@common/constants/Pagination'

export enum SortFields {
  LastUpdatedAt = 'lastModifiedAt',
  AZ09 = 'AZ09',
  ZA90 = 'ZA90',
  Name = 'name'
}

export enum Sort {
  DESC = 'DESC',
  ASC = 'ASC'
}

export interface PageQueryParams {
  page?: number
  size?: number
  searchTerm?: string
  filterIdentifier?: string
  filters?: FilterProperties
}

export type PageQueryParamsWithDefaults = RequiredPick<PageQueryParams, 'page' | 'size'>
export const PAGE_TEMPLATE_DEFAULT_PAGE_SIZE = 10
export const PAGE_TEMPLATE_DEFAULT_PAGE_INDEX = 1

export const usePageQueryParamOptions = (): UseQueryParamsOptions<PageQueryParamsWithDefaults> => {
  const { PL_NEW_PAGE_SIZE } = useFeatureFlags()

  return useQueryParamsOptions({
    page: PAGE_TEMPLATE_DEFAULT_PAGE_INDEX,
    size: PL_NEW_PAGE_SIZE ? COMMON_DEFAULT_PAGE_SIZE : PAGE_TEMPLATE_DEFAULT_PAGE_SIZE
  })
}
