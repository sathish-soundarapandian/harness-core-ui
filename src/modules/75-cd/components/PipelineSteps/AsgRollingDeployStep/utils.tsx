/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import * as Yup from 'yup'
import { getMultiTypeFromValue, MultiTypeInputType } from '@harness/uicore'
import type { UseStringsReturn } from 'framework/strings'

export const minimumHealthyPercentageSchema = (
  getString: UseStringsReturn['getString']
): Yup.Schema<number | undefined> => {
  return Yup.lazy((value: number): any => {
    /* istanbul ignore else */
    if (getMultiTypeFromValue(value as any) === MultiTypeInputType.FIXED) {
      return Yup.mixed().test({
        test(): boolean | Yup.ValidationError {
          if (!value) {
            return true
          }
          if (value > 100) {
            return this.createError({ message: getString('cd.minimumHealthyPercentageMaxLimit') })
          }
          if (value < 0) {
            return this.createError({ message: getString('cd.minimumHealthyPercentageMinLimit') })
          }
          return true
        }
      })
    }
    return Yup.number()
  })
}

export const instanceWarmupSchema = (getString: UseStringsReturn['getString']): Yup.Schema<number | undefined> => {
  return Yup.lazy((value: number): any => {
    /* istanbul ignore else */
    if (getMultiTypeFromValue(value as any) === MultiTypeInputType.FIXED) {
      return Yup.mixed().test({
        test(): boolean | Yup.ValidationError {
          if (!value) {
            return true
          }
          if (value < 0) {
            return this.createError({ message: getString('cd.instanceWarmupError') })
          }
          return true
        }
      })
    }
    return Yup.number()
  })
}
