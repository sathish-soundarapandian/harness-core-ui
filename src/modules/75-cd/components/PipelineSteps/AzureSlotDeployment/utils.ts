import { defaultTo } from 'lodash-es'
import { RUNTIME_INPUT_VALUE, MultiTypeInputType } from '@harness/uicore'
import type { StageElementWrapper } from '@pipeline/utils/pipelineTypes'
import type { StageElementConfig } from 'services/pipeline-ng'
import type { InfraStructureDefinitionYaml, DeploymentStageConfig, EnvironmentYamlV2 } from 'services/cd-ng'

import { getStageFromPipeline } from '@pipeline/components/PipelineStudio/StepUtil'

interface EnvironmentYamlV2Runtime extends Omit<EnvironmentYamlV2, 'infrastructureDefinitions'> {
  infrastructureDefinitions?: InfraStructureDefinitionYaml[] | string
}

interface DeploymentStageConfigAzureWebName extends Omit<DeploymentStageConfig, 'environment'> {
  environment: EnvironmentYamlV2Runtime
}

export interface DeploymentStageElementConfigAzureWebName extends Omit<StageElementConfig, 'spec'> {
  spec?: DeploymentStageConfigAzureWebName
}
export type SelectedStageType = StageElementWrapper<DeploymentStageElementConfigAzureWebName>

export const isRuntimeEnvId = (selectedStage: SelectedStageType): boolean => {
  const envId = selectedStage?.stage?.spec?.environment?.environmentRef
  return !!(envId && envId === RUNTIME_INPUT_VALUE)
}

export const isRuntimeInfraId = (selectedStage: SelectedStageType): boolean => {
  const infraId = selectedStage?.stage?.spec?.environment?.infrastructureDefinitions
  return infraId === RUNTIME_INPUT_VALUE
}

export const isMultiEnv = (selectedStage: SelectedStageType): boolean => {
  const isEnvs = !!selectedStage?.stage?.spec?.environments || !!selectedStage?.stage?.spec?.environmentGroup
  return !!isEnvs
}

export const getEnvId = (selectedStage: SelectedStageType): string => {
  if (isMultiEnv(selectedStage || isRuntimeEnvId(selectedStage))) {
    return ''
  }

  return defaultTo(selectedStage?.stage?.spec?.environment?.environmentRef, '')
}

export const getInfraId = (selectedStage: SelectedStageType): string => {
  const envId = getEnvId(selectedStage)
  if (!envId || isMultiEnv(selectedStage)) {
    return ''
  }

  if (!isRuntimeEnvId(selectedStage) && !isRuntimeInfraId(selectedStage)) {
    const [infra] = selectedStage?.stage?.spec?.environment?.infrastructureDefinitions as InfraStructureDefinitionYaml[]
    return infra.identifier
  }

  return ''
}

export const getEnvIdRuntime = (stageId: string, values: any) => {
  const stage = getStageFromPipeline(stageId, values) as SelectedStageType
  const envId = stage?.stage?.spec?.environment?.environmentRef
  return envId
}

export const getInfraIdRuntime = (stageId: string, values: any): string => {
  const stage = getStageFromPipeline(stageId, values) as SelectedStageType

  if (stage?.stage?.spec?.environment?.infrastructureDefinitions?.[0]) {
    const [{ identifier }] = stage?.stage?.spec?.environment
      ?.infrastructureDefinitions as InfraStructureDefinitionYaml[]
    return identifier
  }
  return ''
}

export const getAllowableTypes = (selectedStage: SelectedStageType): MultiTypeInputType[] => {
  if (isMultiEnv(selectedStage)) {
    return [MultiTypeInputType.EXPRESSION]
  }

  return [MultiTypeInputType.FIXED, MultiTypeInputType.RUNTIME, MultiTypeInputType.EXPRESSION]
}
