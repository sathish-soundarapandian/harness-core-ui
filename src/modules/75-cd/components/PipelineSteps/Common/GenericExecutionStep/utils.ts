/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { FormikErrors, yupToFormErrors } from 'formik'
import * as Yup from 'yup'
import { getMultiTypeFromValue, MultiTypeInputType } from '@harness/uicore'

import type { StepElementConfig } from 'services/cd-ng'
import { getDurationValidationSchema } from '@common/components/MultiTypeDuration/MultiTypeDuration'
import { ValidateInputSetProps, StepViewType } from '@pipeline/components/AbstractSteps/Step'

export const validateGenericFields = ({
  data,
  template,
  getString,
  viewType
}: ValidateInputSetProps<StepElementConfig>): FormikErrors<StepElementConfig> => {
  const isRequired = viewType === StepViewType.DeploymentForm || viewType === StepViewType.TriggerForm
  const errors: FormikErrors<StepElementConfig> = {}

  if (getMultiTypeFromValue(template?.timeout) === MultiTypeInputType.RUNTIME) {
    let timeoutSchema = getDurationValidationSchema({ minimum: '10s' })
    if (isRequired) {
      timeoutSchema = timeoutSchema.required(getString?.('validation.timeout10SecMinimum'))
    }
    const timeout = Yup.object().shape({
      timeout: timeoutSchema
    })
    try {
      timeout.validateSync(data)
    } catch (e) {
      if (e instanceof Yup.ValidationError) {
        const err = yupToFormErrors(e)
        Object.assign(errors, err)
      }
    }
  }
  return errors
}
