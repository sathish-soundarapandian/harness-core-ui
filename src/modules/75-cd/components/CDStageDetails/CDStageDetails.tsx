/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import cx from 'classnames'
import { defaultTo, get } from 'lodash-es'
import { Link, useParams } from 'react-router-dom'
import { PopoverInteractionKind, Position } from '@blueprintjs/core'
import { Popover, Text } from '@harness/uicore'

import { String as StrTemplate } from 'framework/strings'
import type { Application, GitOpsExecutionSummary } from 'services/cd-ng'
import routes from '@common/RouteDefinitions'
import { getScopeFromValue } from '@common/components/EntityReference/EntityReference'
import { getIdentifierFromScopedRef } from '@common/utils/utils'

import { Scope } from '@common/interfaces/SecretsInterface'
import type { ProjectPathProps, ModulePathParams, Module } from '@common/interfaces/RouteInterfaces'
import type { StageDetailProps } from '@pipeline/factories/ExecutionFactory/types'
import { ServicePopoverCard } from '@cd/components/ServicePopoverCard/ServicePopoverCard'
import serviceCardCSS from '@cd/components/ServicePopoverCard/ServicePopoverCard.module.scss'
import css from './CDStageDetails.module.scss'

const GitopsApplications = ({
  gitOpsApps,
  orgIdentifier,
  projectIdentifier,
  accountId,
  module
}: {
  gitOpsApps: Application[]
  orgIdentifier: string
  projectIdentifier: string
  accountId: string
  module: Module
}): React.ReactElement | null => {
  if (gitOpsApps.length === 0) return null

  const firstApp: Application = gitOpsApps[0]

  return (
    <div data-test-id="GitopsApplications">
      <StrTemplate className={css.title} tagName="div" stringID="applications" />
      <ul className={css.values}>
        <li className={css.gitOpsAppsLi}>
          <Link
            onClick={e => e.stopPropagation()}
            to={routes.toGitOpsApplication({
              orgIdentifier,
              projectIdentifier,
              accountId,
              module,
              applicationId: (firstApp.identifier || firstApp.name) as string,
              agentId: firstApp.agentIdentifier
            })}
          >
            <Text>{firstApp.name}</Text>
          </Link>
          {gitOpsApps.length > 1 ? (
            <Popover
              interactionKind={PopoverInteractionKind.HOVER}
              className={css.serviceWrapper}
              position={Position.BOTTOM_RIGHT}
            >
              <span>,&nbsp;+{Math.abs(gitOpsApps.length - 1)}</span>
              <div className={serviceCardCSS.main}>
                <ul className={css.values}>
                  {gitOpsApps.slice(1).map((app: Application, index: number) => {
                    return (
                      <li key={app.identifier || index}>
                        <Link
                          onClick={e => e.stopPropagation()}
                          to={routes.toGitOpsApplication({
                            orgIdentifier,
                            projectIdentifier,
                            accountId,
                            module,
                            applicationId: (app.identifier || app.name) as string,
                            agentId: app.agentIdentifier
                          })}
                        >
                          <Text>{app.name}</Text>
                        </Link>
                      </li>
                    )
                  })}
                </ul>
              </div>
            </Popover>
          ) : null}
        </li>
      </ul>
    </div>
  )
}

export function CDStageDetails(props: StageDetailProps): React.ReactElement {
  const { stage } = props
  const gitOpsApps = get(stage, 'moduleInfo.cd.gitOpsAppSummary.applications') || []
  const { orgIdentifier, projectIdentifier, accountId, module } = useParams<ProjectPathProps & ModulePathParams>()

  const gitOpsEnvironments = Array.isArray(get(stage, 'moduleInfo.cd.gitopsExecutionSummary.environments'))
    ? (get(stage, 'moduleInfo.cd.gitopsExecutionSummary') as Required<GitOpsExecutionSummary>).environments.map(
        envForGitOps =>
          defaultTo({ name: envForGitOps.name, identifier: envForGitOps.identifier }, { name: '', identifier: '' })
      )
    : []
  const serviceScope = getScopeFromValue(get(stage, 'moduleInfo.cd.serviceInfo.identifier', ''))
  const infrastructureScope = getScopeFromValue(get(stage, 'moduleInfo.cd.infraExecutionSummary.identifier', ''))

  return (
    <div className={css.container}>
      <div className={cx(css.main, { [css.threeSections]: !!gitOpsApps.length })}>
        <div>
          <StrTemplate className={css.title} tagName="div" stringID="serviceOrServices" />
          <ul className={css.values}>
            <li>
              <Popover
                wrapperTagName="div"
                targetTagName="div"
                interactionKind="hover"
                position={Position.BOTTOM_RIGHT}
                className={css.serviceWrapper}
              >
                <div className={css.serviceName} data-testid={'serviceLink'}>
                  <Link
                    to={`${routes.toServiceStudio({
                      accountId,
                      ...(serviceScope != Scope.ACCOUNT && { orgIdentifier: orgIdentifier }),
                      ...(serviceScope === Scope.PROJECT && { projectIdentifier: projectIdentifier }),
                      serviceId: getIdentifierFromScopedRef(get(stage, 'moduleInfo.cd.serviceInfo.identifier', '')),
                      module,
                      accountRoutePlacement: 'settings'
                    })}`}
                  >
                    <Text className={css.stageItemDetails} lineClamp={1}>
                      {get(stage, 'moduleInfo.cd.serviceInfo.displayName', null)}
                    </Text>
                  </Link>
                </div>
                <ServicePopoverCard service={get(stage, 'moduleInfo.cd.serviceInfo')} />
              </Popover>
            </li>
          </ul>
        </div>
        <div>
          <StrTemplate className={css.title} tagName="div" stringID="environmentOrEnvironments" />
          <ul className={css.values} data-testid={'environmentLink'}>
            {gitOpsEnvironments.length ? (
              <Text lineClamp={2} className={css.gitOpsEnvText}>
                {gitOpsEnvironments.map(env => {
                  const gitOpsEnvironmentScope = getScopeFromValue(defaultTo(env.identifier, ''))
                  return (
                    <Link
                      key={env.identifier}
                      to={`${routes.toEnvironmentDetails({
                        accountId,
                        ...(gitOpsEnvironmentScope != Scope.ACCOUNT && { orgIdentifier: orgIdentifier }),
                        ...(gitOpsEnvironmentScope === Scope.PROJECT && { projectIdentifier: projectIdentifier }),
                        environmentIdentifier: getIdentifierFromScopedRef(defaultTo(env.identifier, '')),
                        module,
                        accountRoutePlacement: 'settings'
                      })}`}
                    >
                      <Text lineClamp={1} className={css.stageItemDetails} margin={{ right: 'xsmall' }}>
                        {env.name}
                      </Text>
                    </Link>
                  )
                })}
              </Text>
            ) : (
              <Link
                to={routes.toEnvironmentDetails({
                  accountId,
                  ...(infrastructureScope != Scope.ACCOUNT && { orgIdentifier: orgIdentifier }),
                  ...(infrastructureScope === Scope.PROJECT && { projectIdentifier: projectIdentifier }),
                  environmentIdentifier: getIdentifierFromScopedRef(
                    get(stage, 'moduleInfo.cd.infraExecutionSummary.identifier', '')
                  ),
                  module,
                  sectionId: 'INFRASTRUCTURE',
                  accountRoutePlacement: 'settings'
                })}
              >
                <li>
                  <Text lineClamp={1} className={css.stageItemDetails}>
                    {get(stage, 'moduleInfo.cd.infraExecutionSummary.name', null)}
                  </Text>
                </li>
              </Link>
            )}
          </ul>
        </div>
        <GitopsApplications
          gitOpsApps={gitOpsApps}
          orgIdentifier={orgIdentifier}
          projectIdentifier={projectIdentifier}
          accountId={accountId}
          module={module}
        />
      </div>
    </div>
  )
}
