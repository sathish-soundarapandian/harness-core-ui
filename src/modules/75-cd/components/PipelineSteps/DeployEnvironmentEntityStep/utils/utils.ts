/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { RUNTIME_INPUT_VALUE } from '@harness/uicore'
import { isEmpty, isEqual, isNil } from 'lodash-es'
import * as Yup from 'yup'
import type { UseStringsReturn } from 'framework/strings'
import { isValueRuntimeInput } from '@common/utils/utils'
import type {
  DeployEnvironmentEntityConfig,
  DeployEnvironmentEntityCustomStepProps,
  DeployEnvironmentEntityFormState
} from '../types'
import {
  processEnvironmentGroupInitialValues,
  processMultiEnvironmentInitialValues,
  processSingleEnvironmentGitOpsInitialValues,
  processSingleEnvironmentInitialValues
} from './processInitialValuesUtils'
import {
  processEnvironmentGroupFormValues,
  processMultiEnvironmentFormValues,
  processSingleEnvironmentFormValues,
  processSingleEnvironmentGitOpsFormValues
} from './processFormValuesUtils'

export function processInitialValues(
  initialValues: DeployEnvironmentEntityConfig,
  customStepProps: DeployEnvironmentEntityCustomStepProps,
  onUpdate?: (data: DeployEnvironmentEntityConfig) => void
): DeployEnvironmentEntityFormState {
  const gitOpsEnabled = !!customStepProps.gitOpsEnabled

  // istanbul ignore else
  if (initialValues.environment && initialValues.environment.environmentRef) {
    if (gitOpsEnabled) {
      return processSingleEnvironmentGitOpsInitialValues(initialValues.environment)
    } else {
      // this is backwards compatibility for single env multiple clusters in the older redesign setup
      if (
        Array.isArray(initialValues.environment.gitOpsClusters) &&
        initialValues.environment.gitOpsClusters.length > 1
      ) {
        return processMultiEnvironmentInitialValues(
          {
            environments: {
              metadata: {
                parallel: false
              },
              values: [{ ...initialValues.environment }]
            }
          },
          customStepProps
        )
      }
      return processSingleEnvironmentInitialValues(initialValues.environment, customStepProps)
    }
  } else if (initialValues.environments) {
    return processMultiEnvironmentInitialValues(initialValues, customStepProps)
  } else if (initialValues.environmentGroup) {
    // This handles the migration from runtime environment group of old pipeline to post multi infra handling
    if (
      isEqual(initialValues.environmentGroup, {
        envGroupRef: RUNTIME_INPUT_VALUE,
        deployToAll: true
      })
    ) {
      const updatedInitialValues = {
        environmentGroup: {
          envGroupRef: RUNTIME_INPUT_VALUE,
          deployToAll: RUNTIME_INPUT_VALUE as any,
          environments: RUNTIME_INPUT_VALUE as any
        }
      }

      onUpdate?.(updatedInitialValues)
      return processEnvironmentGroupInitialValues(updatedInitialValues, customStepProps)
    } else if (
      // This handles the migration from 'all' environments under environment group of old pipeline to post multi infra handling
      initialValues.environmentGroup.deployToAll === true &&
      !isValueRuntimeInput(initialValues.environmentGroup.envGroupRef) &&
      isNil(initialValues.environmentGroup.envGroupRef)
    ) {
      const updatedInitialValues = {
        environmentGroup: {
          envGroupRef: initialValues.environmentGroup.envGroupRef,
          deployToAll: true,
          environments: RUNTIME_INPUT_VALUE as any
        }
      }

      onUpdate?.(updatedInitialValues)
      return processEnvironmentGroupInitialValues(updatedInitialValues, customStepProps)
    }

    return processEnvironmentGroupInitialValues(initialValues, customStepProps)
  }

  return {
    category: 'single'
  }
}

export function processFormValues(
  data: DeployEnvironmentEntityFormState,
  customStepProps: DeployEnvironmentEntityCustomStepProps
): DeployEnvironmentEntityConfig {
  const gitOpsEnabled = !!customStepProps.gitOpsEnabled

  // istanbul ignore else
  if (data.category === 'single') {
    if (gitOpsEnabled) {
      return processSingleEnvironmentGitOpsFormValues(data)
    } else {
      return processSingleEnvironmentFormValues(data, customStepProps)
    }
  } else if (data.category === 'group') {
    return processEnvironmentGroupFormValues(data, customStepProps)
  } else if (data.category === 'multi') {
    return processMultiEnvironmentFormValues(data, customStepProps)
  }
  return {}
}

export function getValidationSchema(getString: UseStringsReturn['getString'], gitOpsEnabled: boolean): Yup.MixedSchema {
  return Yup.mixed()
    .required()
    .test({
      test(valueObj: DeployEnvironmentEntityFormState): boolean | Yup.ValidationError {
        if (valueObj.category === 'single') {
          if (!valueObj.environment) {
            return this.createError({
              path: 'environment',
              message: getString('cd.pipelineSteps.environmentTab.environmentIsRequired')
            })
          }

          if (!gitOpsEnabled && !valueObj.infrastructure && !isValueRuntimeInput(valueObj.environment)) {
            return this.createError({
              path: 'infrastructure',
              message: getString('cd.pipelineSteps.environmentTab.infrastructureIsRequired')
            })
          }

          // To be put back once BE supports multi environments
          // if (gitOpsEnabled && !valueObj.cluster && !isValueRuntimeInput(valueObj.environment)) {
          //   return this.createError({
          //     path: 'cluster',
          //     message: getString('cd.pipelineSteps.environmentTab.clusterIsRequired')
          //   })
          // }
        } else if (valueObj.category === 'multi') {
          if (
            !(Array.isArray(valueObj.environments) && valueObj.environments.length) &&
            !isValueRuntimeInput(valueObj.environments as unknown as string) &&
            isEmpty(valueObj.environmentFilters?.['fixedScenario'])
          ) {
            return this.createError({
              path: 'environments',
              message: getString('cd.pipelineSteps.environmentTab.environmentsAreRequired')
            })
          }
        } else if (valueObj.category === 'group') {
          if (!valueObj.environmentGroup) {
            return this.createError({
              path: 'environmentGroup',
              message: getString('cd.pipelineSteps.environmentTab.environmentGroupIsRequired')
            })
          }
        }

        return true
      }
    })
}
