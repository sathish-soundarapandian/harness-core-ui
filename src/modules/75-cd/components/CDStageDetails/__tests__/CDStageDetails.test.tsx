/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render, within, screen, RenderResult } from '@testing-library/react'

import { get } from 'lodash-es'
import { TestWrapper } from '@common/utils/testUtils'
import { executionPathProps, modulePathProps } from '@common/utils/routeUtils'
import routes from '@common/RouteDefinitions'
import { getScopeFromValue } from '@common/components/EntityReference/EntityReference'
import { getIdentifierFromScopedRef } from '@common/utils/utils'
import { Scope } from '@common/interfaces/SecretsInterface'
import { CDStageDetails } from '../CDStageDetails'
import props from './props.json'
import propsWithGitOpsApps from './propsWithGitOpsApps.json'
import propsAccountSrvcEnv from './propsAccountSrvcEnv.json'

const getModuleParams = (scope: Scope, module = 'cd') => ({
  accountId: 'accountId',
  ...(scope != Scope.ACCOUNT && { orgIdentifier: 'orgIdentifier' }),
  ...(scope === Scope.PROJECT && { projectIdentifier: 'projectIdentifier' }),
  module
})

const renderComponent = (stageProps: any): RenderResult => {
  return render(
    <TestWrapper
      path={routes.toExecutionPipelineView({ ...executionPathProps, ...modulePathProps })}
      pathParams={{
        ...getModuleParams(Scope.PROJECT),
        pipelineIdentifier: 'pipeline',
        executionIdentifier: 'execution',
        source: 'deployments'
      }}
    >
      <CDStageDetails {...stageProps} />
    </TestWrapper>
  )
}

describe('<CDStageDetails /> tests', () => {
  test('snapshot test', () => {
    const { container } = renderComponent(props)

    expect(container).toMatchSnapshot()
    const serviceLinkElement = screen.getByTestId('serviceLink')
    expect(
      within(serviceLinkElement).getByRole('link', {
        name: 'Order'
      })
    ).toHaveAttribute(
      'href',
      routes.toServiceStudio({
        ...getModuleParams(Scope.PROJECT),
        serviceId: 'Order'
      } as any)
    )
    const envLinkElement = screen.getByTestId('environmentLink')
    expect(
      within(envLinkElement).getByRole('link', {
        name: 'Production'
      })
    ).toHaveAttribute(
      'href',
      routes.toEnvironmentDetails({
        ...getModuleParams(Scope.PROJECT),
        environmentIdentifier: 'Production',
        sectionId: 'INFRASTRUCTURE'
      } as any)
    )
  })

  test('account level service environment route test', () => {
    renderComponent(propsAccountSrvcEnv)

    const srvcId = get(propsAccountSrvcEnv, 'stage.moduleInfo.cd.serviceInfo.identifier', '')
    const envId = get(propsAccountSrvcEnv, 'stage.moduleInfo.cd.infraExecutionSummary.identifier', '')
    const serviceScope = getScopeFromValue(srvcId)
    const infrastructureScope = getScopeFromValue(envId)

    const serviceLinkElement = screen.getByTestId('serviceLink')
    expect(
      within(serviceLinkElement).getByRole('link', {
        name: 'Order'
      })
    ).toHaveAttribute(
      'href',
      routes.toServiceStudio({
        ...getModuleParams(serviceScope),
        serviceId: getIdentifierFromScopedRef(srvcId),
        accountRoutePlacement: 'settings'
      } as any)
    )
    const envLinkElement = screen.getByTestId('environmentLink')
    expect(
      within(envLinkElement).getByRole('link', {
        name: 'Production'
      })
    ).toHaveAttribute(
      'href',
      routes.toEnvironmentDetails({
        ...getModuleParams(infrastructureScope),
        environmentIdentifier: getIdentifierFromScopedRef(envId),
        accountRoutePlacement: 'settings',
        sectionId: 'INFRASTRUCTURE'
      } as any)
    )
  })

  test('test gitops apps', () => {
    const { container } = renderComponent(propsWithGitOpsApps)

    const gitOpsAppsNode = container.querySelector('[data-test-id="GitopsApplications"]')
    expect(gitOpsAppsNode).toMatchSnapshot()
  })

  test('test gitops - account level service environment', () => {
    renderComponent(propsWithGitOpsApps)

    const gitOpsEnvironments = get(propsWithGitOpsApps, 'stage.moduleInfo.cd.gitopsExecutionSummary.environments')
    const accountLevelEnvId = gitOpsEnvironments[0].identifier
    const accountLevelEnvScope = getScopeFromValue(accountLevelEnvId)
    const projectLevelEnvId = gitOpsEnvironments[1].identifier
    const projectLevelEnvScope = getScopeFromValue(projectLevelEnvId)

    const envLinkElement = screen.getByTestId('environmentLink')
    expect(
      within(envLinkElement).getByRole('link', {
        name: 'account level env'
      })
    ).toHaveAttribute(
      'href',
      routes.toEnvironmentDetails({
        ...getModuleParams(accountLevelEnvScope),
        environmentIdentifier: getIdentifierFromScopedRef(accountLevelEnvId),
        accountRoutePlacement: 'settings'
      } as any)
    )
    expect(
      within(envLinkElement).getByRole('link', {
        name: 'customEnv'
      })
    ).toHaveAttribute(
      'href',
      routes.toEnvironmentDetails({
        ...getModuleParams(projectLevelEnvScope),
        environmentIdentifier: getIdentifierFromScopedRef(projectLevelEnvId)
      } as any)
    )
  })
})
