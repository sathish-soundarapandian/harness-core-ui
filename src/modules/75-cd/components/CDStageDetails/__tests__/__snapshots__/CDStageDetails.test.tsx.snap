// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CDStageDetails /> tests snapshot test 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="main"
    >
      <div>
        <div
          class="title"
        >
          serviceOrServices
        </div>
        <ul
          class="values"
        >
          <li>
            <div
              class="bp3-popover-wrapper serviceWrapper"
            >
              <div
                class="bp3-popover-target"
              >
                <div
                  class="serviceName"
                  data-testid="serviceLink"
                  tabindex="0"
                >
                  <a
                    href="/account/accountId/cd/orgs/orgIdentifier/projects/projectIdentifier/services/Order"
                  >
                    <p
                      class="StyledProps--font Text--lineclamp StyledProps--main stageItemDetails"
                      style="--text-line-clamp: 1;"
                    >
                      Order
                    </p>
                  </a>
                </div>
              </div>
            </div>
          </li>
        </ul>
      </div>
      <div>
        <div
          class="title"
        >
          environmentOrEnvironments
        </div>
        <ul
          class="values"
          data-testid="environmentLink"
        >
          <a
            href="/account/accountId/cd/orgs/orgIdentifier/projects/projectIdentifier/environments/Production/details?sectionId=INFRASTRUCTURE"
          >
            <li>
              <p
                class="StyledProps--font Text--lineclamp StyledProps--main stageItemDetails"
                style="--text-line-clamp: 1;"
              >
                Production
              </p>
            </li>
          </a>
        </ul>
      </div>
    </div>
  </div>
</div>
`;

exports[`<CDStageDetails /> tests test gitops apps 1`] = `
<div
  data-test-id="GitopsApplications"
>
  <div
    class="title"
  >
    applications
  </div>
  <ul
    class="values"
  >
    <li
      class="gitOpsAppsLi"
    >
      <a
        href="/account/accountId/cd/orgs/orgIdentifier/projects/projectIdentifier/gitops/applications/ID_1?agentId=AGENT_1"
      >
        <p
          class="StyledProps--font StyledProps--main"
        >
          NAME_1
        </p>
      </a>
      <span
        class="bp3-popover-wrapper serviceWrapper"
      >
        <span
          class="bp3-popover-target"
        >
          <span
            class=""
            tabindex="0"
          >
            ,Â +
            2
          </span>
        </span>
      </span>
    </li>
  </ul>
</div>
`;
